usecases/cluster/transactions_write.go

```diff
@@ -68,6 +80,74 @@ func NewTxManager(remote Remote, logger logrus.FieldLogger) *TxManager {
 	}
 }
 
+func (c *TxManager) resetTxExpiry(ttl time.Duration, id string) {
+	cancel := func() {}
+	ctx := context.Background()
+	if ttl == 0 {
+		c.currentTransactionContext = context.Background()
+	} else {
+		ctx, cancel = context.WithTimeout(ctx, ttl)
+		c.currentTransactionContext = ctx
+	}
+
+	// to prevent a goroutine leak for the new routine we're spawning here,
+	// register a way to terminate it in case the explicit cancel is called
+	// before the context's done channel fires.
+	clearCancelListener := make(chan struct{}, 1)
+
+	c.clearTransaction = func() {
+		c.currentTransaction = nil
+		c.currentTransactionContext = nil
+		c.clearTransaction = func() {}
+
+		clearCancelListener <- struct{}{}
+		close(clearCancelListener)
+	}
+
+	go func(id string) {
+		ctxDone := ctx.Done()
+		select {
```

Comment:
in some rare cases context may expire roughly in the same time `clearTransaction` is called. It is not known which of the cases will be selected here, as both channels will be ready to read. We may end up with expired id added to expired list, when in fact transaction succeeded.
Same problem occurred in `CycleManager::StopAndWait()`

Reply:
Great point. I think in this case we can ignore it. The only unintended side-effect is that an id would be added to the expired list. This means any incoming request for this id would receive the "expired" error message. However, since the TX was already successfully closed, no one should be sending any more requests with this id anyway.